plugin/path_login.go

```diff
@@ -132,91 +134,141 @@ func (b *GcpAuthBackend) parseInfoFromJwt(req *logical.Request, data *framework.
-		return nil, errors.New("provided JWT must have 'kid' header value")
+	key, err := b.getSigningKey(jwtVal, signedJwt.(string), loginInfo.Role, req.Storage)
+	if err != nil {
+		return nil, err
 	}
 
-	// Parse claims
-	loginInfo.JWT, err = jws.ParseJWT(signedJwtBytes)
+	// Parse claims and verify signature.
+	claims := &jwt.Claims{}
+	gceMetadata := &util.GCEIdentityMetadata{}
 	if err != nil {
 		return nil, err
 	}
 
-	sub, ok := loginInfo.JWT.Claims().Subject()
-	if !ok {
-		return nil, errors.New("expected JWT to have 'sub' claim with service account id or email")
+	if err = jwtVal.Claims(key, claims, gceMetadata); err != nil {
+		return nil, err
+	}
+
+	if err = validateJWTClaims(claims, loginInfo.RoleName); err != nil {
+		return nil, err
+	}
+
+	loginInfo.JWTClaims = claims
+	if len(claims.Subject) == 0 {
+		return nil, errors.New("expected JWT to have non-empty 'sub' claim")
+	}
+	loginInfo.ServiceAccountId = claims.Subject
+
+	if len(gceMetadata.InstanceId) > 0 {
+		loginInfo.GceMetadata = gceMetadata
+	}
+
+	if loginInfo.Role.RoleType == gceRoleType && loginInfo.GceMetadata == nil {
+		return nil, errors.New("expected JWT to have claims with GCE metadata")
 	}
-	loginInfo.ServiceAccountId = sub
 
 	return loginInfo, nil
 }
 
-func (info *gcpLoginInfo) validateJWT(keyPEM string, loginInfo *gcpLoginInfo) error {
-	pubKey, err := util.PublicKey(keyPEM)
-	if err != nil {
-		return err
+func (b *GcpAuthBackend) getSigningKey(token *jwt.JSONWebToken, rawToken string, role *gcpRole, s logical.Storage) (interface{}, error) {
+	if len(token.Headers) != 1 {
+		return nil, errors.New("expected token to have exactly one header")
 	}
 
-	validator := &jwt.Validator{
-		Expected: jwt.Claims{
-			"aud": fmt.Sprintf(expectedJwtAudTemplate, loginInfo.RoleName),
-		},
-		Fn: func(c jwt.Claims) error {
-			exp, ok := c.Expiration()
-			if !ok {
-				return errors.New("JWT claim 'exp' is required")
-			}
-			if exp.After(time.Now().Add(loginInfo.Role.MaxJwtExp)) {
-				return fmt.Errorf("JWT expires in %v minutes but must expire within %v for this role. Please generate a new token with a valid expiration.",
-					int(exp.Sub(time.Now())/time.Minute), loginInfo.Role.MaxJwtExp)
-			}
+	keyId := token.Headers[0].KeyID
 
-			return nil
-		},
+	switch role.RoleType {
+	case iamRoleType:
+		iamClient, err := b.IAM(s)
+		if err != nil {
+			return nil, err
+		}
+
+		serviceAccountId, err := util.ParseServiceAccountFromIAMJWT(rawToken)
+		if err != nil {
+			return nil, err
+		}
+
+		accountKey, err := util.ServiceAccountKey(iamClient, keyId, serviceAccountId, role.ProjectId)
+		if err != nil {
+			return nil, err
+		}
+
+		return util.PublicKey(accountKey.PublicKeyData)
+	default:
+		var certsEndpoint string
+		conf, err := b.config(s)
+		if err != nil {
+			return nil, fmt.Errorf("could not read config for backend: %v", err)
+		}
+		if conf != nil {
+			certsEndpoint = conf.GoogleCertsEndpoint
+		}
+
+		key, err := util.Oauth2RSAPublicKey(keyId, certsEndpoint)
```

Comment:
Another minor nitpick: as long as you're camelcasing/properly capitalizing, note that it's OAuth, not Oauth :-)

Reply:
I keep doing that! Will fix